### PR TITLE
[onert/python] Add typing annotations across training API modules

### PR DESCRIPTION
**Pull request by ragmani**
This commit adds typing annotations across training API modules.
- Applied `from typing` imports and added type hints to DataLoader methods and attributes
- Annotated LossFunction, MeanSquaredError, CategoricalCrossentropy constructors with `Literal`, `Dict` and return types
- Typed LossRegistry and MetricsRegistry: `create_*` and `map_*` methods use `str`, `Type[...]`, and enum return types
- Annotated Optimizer base class, Adam, SGD and OptimizerRegistry with `float`, `Optional`, `Dict`, and `Type[...]`
- Added precise types to TrainSession: `train_info: traininfo`, `optimizer: Optional[Optimizer]`, `loss: Optional[LossFunction]`, `metrics: List[Metric]`, `total_time: Dict[str, Union[float,List[float]]]`, and method signatures

ONE-DCO-1.0-Signed-off-by: ragmani <ragmani0216@gmail.com>

### CODE REVIEW

File: runtime/onert/api/python/package/experimental/train/session.py [hunk 1]
```diff
@@ -1,304 +1,276 @@
+import time
 import numpy as np
+from typing import Any, List, Tuple, Dict, Union, Optional
 
 from onert.native import libnnfw_api_pybind
-from onert.native.libnnfw_api_pybind import optimizer as optimizer_type
-from onert.native.libnnfw_api_pybind import loss as loss_type
+# from onert.native.libnnfw_api_pybind import optimizer as optimizer_type, loss as loss_type
```
Comment (zetwhite) [inline runtime/onert/api/python/package/experimental/train/session.py:6]: Maybe.. forgot to remove the commented line? 

```suggestion
```

Comment (ragmani) [reply runtime/onert/api/python/package/experimental/train/session.py]: I removed it.

**Review by chunseoklee — APPROVED** [review]

**Review by seockho-kim — APPROVED** [review]
LGTM

**Review by zetwhite — APPROVED** [review]
